### ISSUE TITLE
A memory leaks if I create strings in a cycle and then these string are collected with GC

### ISSUE THREAD

**Issue by Kasen** [tarantool/tarantool#3890]
Tarantool version:
1.9.2-7-g3736c3798

OS version:
Mac OS 10.14.1

Bug description:
I run a simple script and in two hours Tarantool spends 10 gigabytes of memory of my mac.

Steps to reproduce:
1. Run this script:
```
while true do for _, l in pairs(string.split('qwertyuiopasdfghjklzxcvbnm')) do for i = 1,1024*1024 do local k = string.rep(l, i); fiber.yield() end end end
```
2. wait a couple of hours
3. check memory

**Comment by Totktonada** [comment]
There was the memory fragmentation issue in luajit (see [1] and [2]), but it was fixed a long time ago. So it is not the case.

I tested for the lua memory difference with RSS of the process and it is always aroud 20 MiB (at start as well as after 3.5 hours of `string.rep`'s). It is so as when you call `collectgarbage()` periodically as well as when you don't. The script is attached (it is messy, just to test the case).

So there are no memory leaks (memory that luajit does not count, but RSS does) for the case above.

There are two problem we can consider about:

1. Without explicit `collectgarbage()` call the amount of taken lua memory increases more and more.
2. When the limit of available lua memory is reached (2 GiB AFAIR), tarantool says 'not enough memory' and can stop processing incoming requests instead of aggressively collect garbage.

The first one is not trackerized AFAIK. The second one is pointed by #843 (#562 also seems to be related). I think we should clarify the problem(s) we want to fix and file clear issues/tasks. This is subject to discussion btw team leaders, I think.

So I cast @kostja, @kyukhin and @knazarov here.

[gh-3890-leak.zip](https://github.com/tarantool/tarantool/files/2699618/gh-3890-leak.zip)

[1]: https://www.freelists.org/post/luajit/LuaJIT-memory-fragmentation
[2]: https://github.com/LuaJIT/LuaJIT/commit/d61e9e5d895fcf9443bd02a7a8dedcd41dd700a6
